### DISCUSSION TITLE
How to whitelist custom modules?

### DISCUSSION THREAD

**Discussion by xoxys** [ansible/ansible-lint#1267] [Q&A]
I found some closed issues regarding this topic but couldn't found a sufficient solution.

https://github.com/ansible-community/ansible-lint/issues/245
https://github.com/ansible-community/ansible-lint/issues/372

Is there a way to whitelist custom modules without installing collections/bundle modules? I would like to have a way to just add a list of custom module names to the config.

**Comment by xoxys** [comment]
@ssbarnea Can we transfer the approach from https://github.com/thegeeklab/ansible-later/blob/main/ansiblelater/utils/yamlhelper.py#L414-L417 and extend BUILTIN_TASKS by a user defined list from the Ansible lint config?

**Reply by ssbarnea** [reply]
This is not possible because we run ansible-playbook --syntax-check on playbooks before doing anything else. If that fails we no longer do our own verifications.

Shortly, the only way to make this work is to make Ansible itself believe these modules exists. At some point I was even contemplating the idea of generating module stubs for whitelisted modules in order to please Ansible.

Keep in mind that we run ansible syntax check as subprocess, so these fake modules really need to be added manually.

The reason why I did not spent time implementing the stubs is that it does add a lot of magic that does diverge a lot from production. The fact that Ansible failst find these module it means that there is a high chance for this to happen on production too.

Also, disabling this prevents us from doing better testing later, like validating module args.

If you are curious, I am aware of one place where stubs were used to bypass missing modules: https://opendev.org/zuul/zuul-jobs/src/branch/master/tests/fake-ansible/library

**Reply by xoxys** [reply]
Thanks for the explanation. While I understand the backgrounds, I need to think about it a bit more... It adds a lot of complexity, especially to the user setup, and right now I have no clue how to get it working. `ansible-galaxy collection list` shows the required collection as installed but ansible-lint still complains about an unknown module. How to debug such issues?

**Reply by ssbarnea** [reply]
In order not to mess user installed roles/collections, Ansible lint uses a temporary alternative locations for them: `.cache/[roles|collections]`.

If it finds a `requirements.yml` file, it will try to install it inside `.cache`. So basically you need to add this file an let it install them.
